File: Data-Structures-and-Algorithms/0104-MaximumDepthofBinaryTree/maxDepth.js
function TreeNode(val, left, right) {
    this.val = (val===undefined ? 0 : val)
    this.left = (left===undefined ? null : left)
    this.right = (right===undefined ? null : right)
}
 
const maxDepth = function(root) {
    if (!root) {
        return 0;
    }
    let left = maxDepth(root.left);
    let right = maxDepth(root.right);
    return Math.max(left, right) + 1;
};


let root = new TreeNode(0);
let one = new TreeNode(1);
let two = new TreeNode(2);
let three = new TreeNode(3);
root.left = one;
root.right = two;
two.right = three;
console.log(maxDepth(root))

// O(N) where N is the total number of nodes
// O(N) for worst case and O(log2(N)) for best case
// Max depth of a binary tree

File: Data-Structures-and-Algorithms/0104-MaximumDepthofBinaryTree/maxDepth.py
class TreeNode(object):
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

class Solution(object):
    def maxDepth(self, root):
        if not root:
            return 0
        
        stack = [(root, 1)]
        ans = 0
        while stack:
            node, level = stack.pop()
            ans = max(ans, level)
            if node.left:
                stack.append((node.left, level+1))
            if node.right:
                stack.append((node.right, level+1))
                
        return ans

root  = TreeNode(0)
one   = TreeNode(1)
two   = TreeNode(2)
three = TreeNode(3)
four = TreeNode(4)
root.left = one
root.right = two
two.right = three
three.right = four
print(Solution().maxDepth(root))

# O(N) where N is the total number of nodes
# O(N) for worst case and O(log2(N)) for best case
# Max depth of a binary tree